src/handler.js

```diff
@@ -0,0 +1,330 @@
+import parseRoute from './util/route_parser';
+
+const TASK_TREEHERDER_SCHEMA = 'http://schemas.taskcluster.net/taskcluster-treeherder/v1/task-treeherder-config.json#';
+
+function stateFromRun(run) {
+  switch (run.state) {
+    case 'exception':
+    case 'failed':
+      return 'completed';
+    default:
+      return run.state;
+  }
+}
+
+function resultFromRun(run) {
+  switch (run.state) {
+    case 'completed':
+      return 'success';
+    case 'failed':
+      return 'fail';
+    case 'exception':
+      if (run.reasonResolved === 'canceled') {
+        return 'canceled';
+      }
+
+      return 'exception';
+    default:
+      return 'unknown';
+  }
+}
+
+function taskIsRerun(runId, run) {
+  // TODO: add array includes plugin to tc-rules and use here instead of indexOf
+  return runId > 0 && ['rerun', 'retry'].indexOf(run.reasonCreated) !== -1;
+}
+
+// Creates a log entry for Treeherder to retrieve and parse.  This log is
+// displayed on the Treeherder Log Viewer once parsed.
+function createLogReference(queue, taskId, run) {
+  let logUrl = queue.buildUrl(
+    queue.getArtifact, taskId, run.runId, 'public/logs/live_backing.log'
+  );
+
+  return {
+    // XXX: This is a magical name see 1147958 which enables the log viewer.
+    name: 'builds-4h',
+    url: logUrl
+  };
+}
+
+// Filters the task routes for the treeherder specific route.  Once found,
+// the route is parsed into distinct parts used for constructing the
+// Treeherder job message.
+function parseRouteInfo(prefix, taskId, routes) {
+  let route = routes.filter((r) => {
+    return r.split('.')[0] === prefix;
+  });
+
+  if (route.length != 1) {
+    throw new Error(
+      `Could not determine treeherder route.  Either there is no route, ` +
+      `or more than one matching route exists.  Task ID: ${taskId} Routes: ${routes}`
+    );
+  }
+
+  return parseRoute(route[0]);
+}
+
+
+function validateTask(validate, taskId, task, schema) {
+    if (!task.extra || !task.extra.treeherder) {
+      throw new Error(`Message is missing Treeherder job configuration. Task ID: ${taskId}`);
+    }
+    let validationErrors = validate(task.extra.treeherder, schema);
+    if (validationErrors) {
+      throw new Error(`Message contains an invalid Treeherder job configuration. Task ID: ${taskId} ${validationErrors}`);
+    }
+}
+
+export class Handler extends events.EventEmitter {
+  constructor(options) {
+    super();
+    this.queueEvents = options.queueEvents;
+    this.queue = options.queue;
+    this.scheduler = options.scheduler;
+    this.listener = options.listener;
+    this.prefix = options.prefix;
+    this.publisher = options.publisher;
+    this.validator = options.validator;
+  }
+
+  // Starts up the message handler and listens for messages
+  async start() {
+    this.listener.on('message', async (message) => {
+      await this.handleMessage(message).catch(err => {
+        console.log(`Error caught when processing message. ${err.message}. ${err.stack}`);
```

Comment:
Talk to bstack, about using the monitor thing... you can send this to sentry as a warning...